src/main/java/uk/ac/soton/itinnovation/security/modelvalidator/attackpath/RecommendationsAlgorithm.java

```diff
@@ -0,0 +1,490 @@
+/////////////////////////////////////////////////////////////////////////
+//
+// © University of Southampton IT Innovation Centre, 2023
+//
+// Copyright in this software belongs to University of Southampton
+// IT Innovation Centre of Gamma House, Enterprise Road,
+// Chilworth Science Park, Southampton, SO16 7NS, UK.
+//
+// This software may not be used, sold, licensed, transferred, copied
+// or reproduced in whole or in part in any manner or form or in or
+// on any media by any person other than in accordance with the terms
+// of the Licence Agreement supplied with the software, or otherwise
+// without the prior written consent of the copyright owners.
+//
+// This software is distributed WITHOUT ANY WARRANTY, without even the
+// implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+// PURPOSE, except where stated in the Licence Agreement supplied with
+// the software.
+//
+//      Created By:             Panos Melas
+//      Created Date:           2023-01-24
+//      Created for Project :   Cyberkit4SME
+//
+/////////////////////////////////////////////////////////////////////////
+package uk.ac.soton.itinnovation.security.modelvalidator.attackpath;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.springframework.stereotype.Component;
+
+import uk.ac.soton.itinnovation.security.model.system.RiskCalculationMode;
+import uk.ac.soton.itinnovation.security.model.system.RiskVector;
+import uk.ac.soton.itinnovation.security.modelquerier.IQuerierDB;
+import uk.ac.soton.itinnovation.security.modelquerier.dto.ModelDB;
+import uk.ac.soton.itinnovation.security.modelvalidator.Progress;
+import uk.ac.soton.itinnovation.security.systemmodeller.rest.dto.recommendations.CSGNode;
+import uk.ac.soton.itinnovation.security.systemmodeller.rest.dto.recommendations.ControlDTO;
+import uk.ac.soton.itinnovation.security.systemmodeller.rest.dto.recommendations.ControlStrategyDTO;
+import uk.ac.soton.itinnovation.security.systemmodeller.rest.dto.recommendations.RecommendationDTO;
+import uk.ac.soton.itinnovation.security.systemmodeller.rest.dto.recommendations.RecommendationReportDTO;
+import uk.ac.soton.itinnovation.security.systemmodeller.rest.dto.recommendations.StateDTO;
+
+import com.bpodgursky.jbool_expressions.Expression;
+import com.bpodgursky.jbool_expressions.Variable;
+
+@Component
+public class RecommendationsAlgorithm {
+
+    private static final Logger logger = LoggerFactory.getLogger(RecommendationsAlgorithm.class);
+
+    private AttackPathDataset apd;
+    private IQuerierDB querier;
+    private String modelId;
+    private int recCounter = 0;
+    private RecommendationReportDTO report;
+    private String riskMode = "CURRENT";
+
+    // allPaths flag for single or double backtrace
+    private boolean allPaths = true;
```

Comment:
This is a confusing variable name. According to its usage, when true it means just the shortest path will be considered. If false then all paths will be considered.